src/cryptenroll/cryptenroll-pkcs11.c

```diff
@@ -40,24 +40,78 @@ int enroll_pkcs11(
+                        return log_oom();
+
+                r = genuine_random_bytes(decrypted_key, decrypted_key_size, RANDOM_BLOCK);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to generate random key: %m");
+
+                r = rsa_encrypt_bytes(pkey, decrypted_key, decrypted_key_size, &encrypted_key, &encrypted_key_size);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to encrypt key: %m");
+        }
+        else if (type == EVP_PKEY_EC) {
+                _cleanup_(EVP_PKEY_freep) EVP_PKEY *pkey_new = NULL;
+                _cleanup_(erase_and_freep) uint8_t *shared_secret = NULL;
+                size_t shared_secret_len = 0;
+                EC_KEY *ec_key = EVP_PKEY_get0_EC_KEY(pkey);
+                int nid;
+
+                if (!ec_key)
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "pkey doesn't have EC_KEY associated");
+
+                nid = EC_GROUP_get_curve_name(EC_KEY_get0_group(ec_key));
+                r = pkey_generate_ec_key(nid, &pkey_new);
+                if (r != 0)
+                        return log_error_errno(r, "Failed to generate ec key");
+
+                r = pkey_ecdh_derive_shared_secret(pkey_new, pkey, NULL, &shared_secret_len);
+                if (r != 0 || shared_secret_len == 0)
+                        return log_error_errno(r, "Failed to determine derived shared secret size");
+
+                shared_secret = malloc(shared_secret_len);
+                if (!shared_secret)
+                        return log_oom();
+
+                r = pkey_ecdh_derive_shared_secret(pkey_new, pkey, shared_secret, &shared_secret_len);
+                if (r != 0)
+                        return log_error_errno(r, "Failed to derive shared secret");
+
+                decrypted_key = malloc(72);
+                if (!decrypted_key)
+                        return log_oom();
+
+                r = hkdf_sha512(shared_secret, shared_secret_len, 72, decrypted_key);
+                if (r != 0)
+                        return log_error_errno(r, "Failed to generate hkdf secret");
+
+                decrypted_key_size = 72;
+                encrypted_key_size = i2d_PUBKEY(pkey_new, NULL);
+                if (encrypted_key_size == 0)
+                        return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to determine encoded public key size");
+
+                encrypted_key = malloc(encrypted_key_size);
+                if (!encrypted_key)
+                        return log_oom();
+
+                // i2d_PUBKEY function has a side effect that makes *pp point to end of the allocated buffer
+                uint8_t *buffer = encrypted_key;
+                encrypted_key_size = i2d_PUBKEY(pkey_new, (uint8_t **)&buffer);
+                if (encrypted_key_size == 0)
+                        return  log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to get encoded public key");
+        }
```

Comment:
btw, somewhere here, at the top of the enrollment code it would be good to have a coment explaining what the construction is, i.e. mention ECDH, mention which keys are stored on the PCKS#11 device, which key is stored in the unencrypted LUKS2 metadata header, and which key is simply dumped and forgotten. Also mention that the resulting symmetric key is not expanded here but later by the LUKS2 key management.